lib/App/Config/Chronicle.pm

```diff
@@ -347,6 +350,316 @@ sub current_revision {
+If the third argument is set to 1 the result of the query will be cached in Redis. This is useful if a certain
+    revision will be needed repeatedly, to avoid excessive database access. By default this argument is 0.
+All cached revisions will become stale if the key is set with a new value.
+
+Example:
+    get_history('system.email', 0); Retrieves current version
+    get_history('system.email', 1); Retreives previous revision
+    get_history('system.email', 2); Retreives version before previous
+
+=cut
+
+sub get_history {
+    my ($self, $key, $rev, $cache) = @_;
+    $cache //= 0;
+
+    die "Cannot get history of key: $key | Key must be dynamic" unless $self->_key_is_dynamic($key);
+
+    my @data_objs = $self->_retrieve_objects([$key, $key . '::' . $rev]);
+    my $curr_obj  = $data_objs[0];
+    my $hist_obj  = $data_objs[1];
+
+    # If no cache, or cache is stale, get from db
+    unless ($hist_obj && $hist_obj->{_local_rev} == $curr_obj->{_local_rev}) {
+        $hist_obj = $self->chronicle_reader->get_history($self->setting_namespace, $key, $rev);
+
+        $hist_obj->{_local_rev} = $curr_obj->{_local_rev} if $hist_obj;
+        $self->_store_objects(
+            {$key . '::' . $rev => $hist_obj},
+            Date::Utility->new,
+            [
+                0,         #<-- IMPORTANT: disables archiving
+                0,         #<-- IMPORTANT: supresses publication
+                604800,    #<-- ttl = 7 days so that stale histories don't stay indefinitely.
+            ]) if $hist_obj && $cache;
+    }
+    return $hist_obj->{data} if $hist_obj;
+    return undef;
+}
+
+=head2 subscribe
+
+Subscribes to changes for the specified $key with the sub $subref called when a new value is set.
+The chronicle_writer must have publish_on_set enabled.
+
+=cut
+
+sub subscribe {
+    my ($self, $key, $subref) = @_;
+    $self->chronicle_writer->subscribe($self->setting_namespace, $key, $subref);
+}
+
+=head2 unsubscribe
+
+Stops the sub $subref from being called when $key is set with a new value.
+The chronicle_writer must have publish_on_set enabled.
+
+=cut
+
+sub unsubscribe {
+    my ($self, $key, $subref) = @_;
+    $self->chronicle_writer->unsubscribe($self->setting_namespace, $key, $subref);
+}
+
+has _keys => (
+    is      => 'ro',
+    isa     => 'HashRef',
+    default => sub { {} },
+);
+
+sub _all_keys {
+    my $self = shift;
+    return keys %{$self->_keys};
+}
+
+sub _dynamic_keys {
+    my $self = shift;
+    return grep { $self->_key_is_dynamic($_) } $self->_all_keys();
+}
+
+sub _static_keys {
+    my $self = shift;
+    return grep { $self->_key_is_static($_) } $self->_all_keys();
+}
+
+sub _key_exists {
+    my ($self, $key) = @_;
+    return any { $key eq $_ } $self->_all_keys();
+}
+
+sub _key_is_dynamic {
+    my ($self, $key) = @_;
+    return unless $self->_key_exists($key);
+    return $self->_keys->{$key}->{key_type} eq 'dynamic';
```

Comment:
If these are supposed to be internal high-performance methods, it would probably be better to make the test itself more efficient. Combine the two operations with something like

    my $keys = $self->_keys;
    return exists $keys->{$key} && $keys->{$key}->{key_type} eq 'dynamic';